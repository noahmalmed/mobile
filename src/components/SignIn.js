import React from 'react'
import {
  Alert,
  Linking,
  ScrollView,
  Text,
  TouchableOpacity,
  View
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import PropTypes from 'prop-types';
import { signIn, continueAsGuest } from '../actions/auth'
import { connect } from 'react-redux';
import Button from './Button'
import Input from './Input'
import NavBar from './NavBar'
import OverlaySpinner from './OverlaySpinner'
import StyledText from './StyledText'
import { Actions } from 'react-native-router-flux'
import { bindActionCreators } from 'redux'
import * as navBarActions from '../actions/navBar'

const mapStateToProps = (state) => ({
  isFetching: state.main.isFetching,
  errorMessage: state.main.errorMessage,
})

const mapDispatchToProps = (dispatch) => ({
  navBarActions: bindActionCreators(navBarActions, dispatch),
  signIn(login, password) {
    dispatch(signIn(login, password))
  },
  continueAsGuest() {
    dispatch(continueAsGuest())
  },
})

const PAGE_KEY = 'SignIn';

export class SignIn extends React.Component {
  constructor(props) {
    super(props)
    this.state = {login: '', password: ''}
    this.handleRegistration = this.handleRegistration.bind(this)
    this.handleResetPassword = this.handleResetPassword.bind(this)
    this.handleSignIn = this.handleSignIn.bind(this)
    this.continueAsGuest = this.continueAsGuest.bind(this)
  }

  componentDidMount() {
    this.props.navBarActions.setNavbarBackButton(this.props.showBackButton, PAGE_KEY)
  }

  handleRegistration() {
    Actions.Register()
  }

  handleResetPassword() {
    const zurl='http://zooniverse.org/reset-password'
    Linking.canOpenURL(zurl).then(supported => {
      if (supported) {
        Linking.openURL(zurl);
      } else {
        Alert.alert(
          'Error',
          'Sorry, but it looks like you are unable to reset your password in your default browser.',
        )
      }
    });
  }

  handleSignIn() {
    //prevent red screen of death thrown by a console.error in javascript-client
    /* eslint-disable no-console */
    console.reportErrorsAsExceptions = false
    this.props.signIn(this.state.login, this.state.password)
  }

  continueAsGuest() {
    this.props.continueAsGuest()
  }

  static renderNavigationBar() {
    return <NavBar pageKey={PAGE_KEY} showLogo={true} showDrawer={false} />;
  }

  render() {
    const signInDisabled = ( (this.state.login === '') || (this.state.password ===  '') ? true : false )

    const errorMessage =
      <StyledText
      textStyle={'errorMessage'}
      text={ this.props.errorMessage } />

    return (
      <View style={styles.container}>
        <ScrollView style={styles.scrollView}>
          <View style={styles.signInContainer}>
            <StyledText
              textStyle={'headerText'}
              text={'SIGN IN'} />
            <Input
              labelText={'Username or Email Address'}
              handleOnChangeText={(login) => this.setState({login})} />
            <Input
              labelText={'Password'}
              passwordField={true}
              handleOnChangeText={(password) => this.setState({password})} />
            { this.props.errorMessage ? errorMessage : null }
            <TouchableOpacity
              onPress={this.handleResetPassword} style={styles.forgotPasswordContainer}>
              <StyledText
                textStyle={'link'}
                text={ 'Forget your password?' } />
            </TouchableOpacity>

            <Button
              handlePress={this.handleSignIn}
              disabled={signInDisabled}
              buttonStyle={ signInDisabled ? 'disabledButton' : null }
              text={'Sign In'} />

            <View style={styles.lined}>
              <View style={styles.lineThrough} />
              <Text style={styles.centerText}>OR</Text>
              <View style={styles.lineThrough} />
            </View>

            <Button
              handlePress={this.continueAsGuest}
              text={'Continue without signing in'} />

            <Button
              handlePress={this.handleRegistration}
              buttonStyle={'registerButton'}
              text={'Register for account'} />
          </View>
        </ScrollView>
        { this.props.isFetching ? <OverlaySpinner /> : null }
      </View>
    );
  }
}

const styles = EStyleSheet.create({
  scrollView: {
    backgroundColor: 'white',
  },
  container: {
    flex: 1,
  },
  signInContainer: {
    backgroundColor: 'transparent',
    margin: 30,
    marginTop: 20,
  },
  lined: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    height: 20,
    marginTop: 20,
    marginBottom: 10
  },
  lineThrough: {
    flex: 1,
    height: 1,
    backgroundColor: 'black'
  },
  centerText: {
    height: 20,
    lineHeight: 20,
    margin: 20
  },
  forgotPasswordContainer: {
    paddingTop: 10,
    paddingBottom: 10
  }
});

SignIn.propTypes = {
  showBackButton: PropTypes.bool,
  isFetching: PropTypes.bool,
  signIn: PropTypes.func,
  continueAsGuest: PropTypes.func,
  errorMessage: PropTypes.string,
  navBarActions: PropTypes.shape({
    setNavbarBackButton: PropTypes.func
  })
}

SignIn.defaultProps = {
  showBackButton: false
}

export default connect(mapStateToProps, mapDispatchToProps)(SignIn)