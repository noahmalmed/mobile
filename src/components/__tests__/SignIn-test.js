import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { SignIn } from '../SignIn'

const navBarActions = {
  setNavbarBackButton: () => {}
}

it('renders correctly', () => {
  const tree = renderer.create(
    <SignIn isConnected={true} navBarActions={navBarActions} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders error message', () => {
  const tree = renderer.create(
    <SignIn navBarActions={navBarActions} isConnected={true} errorMessage={'does not compute'} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
