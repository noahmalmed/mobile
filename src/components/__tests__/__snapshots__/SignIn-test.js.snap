// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={undefined}
>
  <RCTScrollView
    style={undefined}
  >
    <View>
      <View
        style={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": "Karla",
              },
              Array [
                undefined,
                undefined,
              ],
            ]
          }
        >
          SIGN IN
        </Text>
        <View
          style={undefined}
        >
          <View
            style={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "fontFamily": "Karla",
                  },
                  Array [
                    undefined,
                  ],
                ]
              }
            >
              Username or Email Address
            </Text>
          </View>
          <View
            style={undefined}
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              keyboardType="default"
              onChangeText={[Function]}
              returnKeyType="done"
              secureTextEntry={false}
              style={undefined}
              underlineColorAndroid="white"
            />
          </View>
        </View>
        <View
          style={undefined}
        >
          <View
            style={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "fontFamily": "Karla",
                  },
                  Array [
                    undefined,
                  ],
                ]
              }
            >
              Password
            </Text>
          </View>
          <View
            style={undefined}
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              keyboardType="default"
              onChangeText={[Function]}
              returnKeyType="done"
              secureTextEntry={true}
              style={undefined}
              underlineColorAndroid="white"
            />
          </View>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "fontFamily": "Karla",
                },
                Array [
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Forget your password?
          </Text>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}
          >
            Sign In
          </Text>
        </View>
        <View
          style={undefined}
        >
          <View
            style={undefined}
          />
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}
          >
            OR
          </Text>
          <View
            style={undefined}
          />
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}
          >
            Continue without signing in
          </Text>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}
          >
            Register for account
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`renders error message 1`] = `
<View
  style={undefined}
>
  <RCTScrollView
    style={undefined}
  >
    <View>
      <View
        style={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": "Karla",
              },
              Array [
                undefined,
                undefined,
              ],
            ]
          }
        >
          SIGN IN
        </Text>
        <View
          style={undefined}
        >
          <View
            style={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "fontFamily": "Karla",
                  },
                  Array [
                    undefined,
                  ],
                ]
              }
            >
              Username or Email Address
            </Text>
          </View>
          <View
            style={undefined}
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              keyboardType="default"
              onChangeText={[Function]}
              returnKeyType="done"
              secureTextEntry={false}
              style={undefined}
              underlineColorAndroid="white"
            />
          </View>
        </View>
        <View
          style={undefined}
        >
          <View
            style={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "fontFamily": "Karla",
                  },
                  Array [
                    undefined,
                  ],
                ]
              }
            >
              Password
            </Text>
          </View>
          <View
            style={undefined}
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              keyboardType="default"
              onChangeText={[Function]}
              returnKeyType="done"
              secureTextEntry={true}
              style={undefined}
              underlineColorAndroid="white"
            />
          </View>
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "fontFamily": "Karla",
              },
              Array [
                undefined,
                undefined,
              ],
            ]
          }
        >
          does not compute
        </Text>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "fontFamily": "Karla",
                },
                Array [
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Forget your password?
          </Text>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}
          >
            Sign In
          </Text>
        </View>
        <View
          style={undefined}
        >
          <View
            style={undefined}
          />
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}
          >
            OR
          </Text>
          <View
            style={undefined}
          />
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}
          >
            Continue without signing in
          </Text>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}
          >
            Register for account
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
